File: FindWeather/adapter/WeatherStackAdapter.java
package adapter;
import api.WeatherStackAPI;
import models.WeatherData;
import org.json.JSONObject;

public class WeatherStackAdapter implements WeatherProvider{
    private final WeatherStackAPI api;
    public WeatherStackAdapter() {
        this.api = new WeatherStackAPI();
    }
    @Override
    public WeatherData getWeatherByCity(String city) throws Exception {
        JSONObject response = api.getWeatherByCity(city);

        if (!response.has("current")) {
            throw new Exception("Weather data not available for the city: " + city);
        }

        JSONObject current = response.getJSONObject("current");
        double temperature = current.getDouble("temperature");
        String conditions = current.getJSONArray("weather_descriptions").getString(0);

        return new WeatherData(city, temperature, conditions, "WeatherStack");
    }

    @Override
    public WeatherData getWeatherByIP(double latitude, double longitude) throws Exception {
        JSONObject response = api.getWeatherByIP(latitude, longitude);
        double temperature = response.getJSONObject("current").getDouble("temperature");
        String conditions = response.getJSONObject("current").getJSONArray("weather_descriptions").getString(0);

        return new WeatherData("Location via IP", temperature, conditions, "WeatherStack");
    }
}

File: FindWeather/FindWeather/Weather/WeatherStackAdapter.cs
using System.Text.Json;
using FindWeather.Data;

namespace FindWeather.Weather;

public class WeatherStackAdapter : IWeatherAPI
{
    private readonly HttpClient _httpClient;
    private const string ApiKey = "462e316b52668a55c98e4a666087aba5";

    public WeatherStackAdapter()
    {
        _httpClient = new HttpClient();
    }

    public WeatherData GetWeatherData(double latitude, double longitude)
    {
        var response = _httpClient.GetStringAsync(
            $"http://api.weatherstack.com/current?access_key={ApiKey}&query={latitude},{longitude}"
        ).Result;

        using (var jsonDoc = JsonDocument.Parse(response))
        {
            var location = jsonDoc.RootElement.GetProperty("location").GetProperty("name").GetString()!;
            var temperature = jsonDoc.RootElement.GetProperty("current").GetProperty("temperature").GetDouble()!;
            var conditions = jsonDoc.RootElement.GetProperty("current").GetProperty("weather_descriptions")[0].GetString()!;

            return new WeatherData
            {
                Location = location,
                Temperature = temperature,
                Conditions = conditions
            };
        }
    }
}
